src/java.base/share/classes/java/io/Reader.java

```diff
@@ -140,6 +140,205 @@ public void close() {
         };
     }
 
+    /**
+     * Returns a new {@code Reader} whose source is a {@link CharSequence}.
+     *
+     * <p> The resulting reader is not safe for use by multiple
+     * concurrent threads. If the reader is to be used by more than one
+     * thread it should be controlled by appropriate synchronization.
+     *
+     * <p> The returned reader is initially open. The reader is closed by
+     * calling the {@code close()} method. Subsequent calls to {@code close()}
+     * have no effect.
+     *
+     * <p> After the reader has been closed, the {@code read()},
+     * {@code read(char[])}, {@code read(char[], int, int)},
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The {@code markSupported()} method returns {@code true}.
+     *
+     * @param source {@code CharSequence} providing the character stream.
+     *
+     * @return a {@code Reader} which reads characters from {@code source}
+     *
+     * @throws NullPointerException if {@code source} is {@code null}
+     *
+     * @since 24
+     */
+    public static Reader of(CharSequence source) {
+        Objects.requireNonNull(source);
+
+        return new Reader() {
+            private final int length = source.length();
+            private CharSequence cs = source;
```

Comment:
This is correct, but then we would keep the reference to the char sequence "forever", without any good reason. This content could be *huge* (and often it is, as `CharSequence` most often is a `StringBuilder` just because the application *wants* to create huge content, and `String`-concatenation would be too costly with *huge* content). In fact, I see this scenario as the top use case for this new API.

Having said that, I would really like to have this field non-final to unlink the reference to the source as soon as possible, implicitly allowing GC to free this unused memory.

Reply:
> keep the reference to the char sequence "forever"

Note that this only happens if the `Reader` instance is kept forever; but a `Reader` is supposed to be used and closed and discarded instead of being held indefinitely after closing. If users are bad enough, they can never call `close` and the non-final `cs` is still leaked.

The right way of GC freeing is to let the GC free the reader and the char sequence together, instead of allowing to free when a client keeps this reader instance erroneously after closing.

Reply:
I really do see you point, nevertheless I think we should be kind to "bad" users, too. In the end, there is *no* specification which explicitly tells "bad" users that readers are support to be short-living (despite it being rather obvious), while `close` explicitly mentions freeing *all* resources, so it is rather unexpected to keep something past `close`, even for short term. NB: `StringReader` applies the very same behavior, too (using `s` and `str` variables), that's where I copied the code from originally. It is not a big problem for us to have two variables, neither that the name of the variable is `source` (in fact, I do like `source` even more than `cs`, but this is just my personal preference). To sum up, I really dislike the idea to keep the reference for any longer than essentially needed. If it's just about the name of the variable, I could simply switch them if you prefer that.

Reply:
On a linux x64 build:
```
./javap -p java.io.Reader\$2
Compiled from "Reader.java"
class java.io.Reader$2 extends java.io.Reader {
  private final int length;
  private java.lang.CharSequence cs;
  private int next;
  private int mark;
  final java.lang.CharSequence val$source;
  java.io.Reader$2(java.lang.CharSequence);
  private void ensureOpen() throws java.io.IOException;
  public int read() throws java.io.IOException;
  public int read(char[], int, int) throws java.io.IOException;
  public long skip(long) throws java.io.IOException;
  public boolean ready() throws java.io.IOException;
  public boolean markSupported();
  public void mark(int) throws java.io.IOException;
  public void reset() throws java.io.IOException;
  public void close();
}
```

The javap in the built binaries of this patch shows that this anonymous class already keeps the `CharSequence val$source` in a final field; therefore, your setting the `cs` field to `null` will not help GC or anything of that sort.